File: packages/python/replica/starfish_replica/channel.py
"""The pluggable sync seam: :class:`ReplicaChannel`.

Mirrors the TS package's ``channel.ts``. A :class:`ChannelScheduler` (see
``scheduler.py``) drives any number of channels on a schedule; a channel
knows only how to sync itself once, given a call context. The HTTP-pull path
(``http_channel.py``) is one channel implementation; ``space/mirror_channel.py``
is another.
"""

from __future__ import annotations

import asyncio
from dataclasses import dataclass, field
from typing import Literal, Optional, Protocol, runtime_checkable

from starfish_replica.config import SyncTrigger

__all__ = [
    "ReplicaCallContext",
    "REPLICATOR_CTX",
    "ReplicaChannel",
    "ChannelSchedule",
    "ScheduledChannel",
    "SyncTrigger",
]


@dataclass(frozen=True)
class ReplicaCallContext:
    """Distinguishes a replication-driven call from a direct app call, so one
    shared data-access function can serve both call sites and branch on which
    it is.

    ``cancel`` is the Python stand-in for TS's ``AbortSignal`` — optional and
    unused by the HTTP channel; a channel that supports cancellation may check
    it between steps.
    """

    call_kind: Literal["replicator", "classic"] = "replicator"
    cancel: Optional[asyncio.Event] = None


REPLICATOR_CTX = ReplicaCallContext(call_kind="replicator")


@runtime_checkable
class ReplicaChannel(Protocol):
    """A single sync-able data path. Implementations: :class:`HttpReplicaChannel`
    (``http_channel.py``), :class:`SpaceMirrorChannel` (``space/mirror_channel.py``)."""

    name: str

    async def sync(self, ctx: ReplicaCallContext) -> None:
        """Perform one sync cycle. Raise on failure — the scheduler funnels
        exceptions to its ``on_error`` handler; a channel should not swallow
        its own errors."""
        ...


@dataclass
class ChannelSchedule:
    """When/how often a :class:`ReplicaChannel` should be synced."""

    triggers: list[SyncTrigger] = field(default_factory=lambda: [SyncTrigger.SCHEDULED])
    interval_ms: Optional[int] = None
    on_pull_min_interval_ms: Optional[int] = None


@dataclass
class ScheduledChannel:
    """A channel paired with its schedule — the unit :class:`ChannelScheduler` operates on."""

    channel: ReplicaChannel
    schedule: ChannelSchedule

File: packages/ts/replica/src/channel.ts
/**
 * The pluggable data-path seam `ReplicaManager` schedules against.
 *
 * `ReplicaManager` owns ONLY scheduling (interval loop, on_pull cooldown,
 * error funnel) — it knows nothing about HTTP, local storage, or Starfish
 * spaces. Each `ReplicaChannel` owns one collection's actual sync mechanics
 * behind a single `sync()` seam. `HttpReplicaChannel` (../http-channel.ts) is
 * the default, unchanged primary→replica-server path; `SpaceMirrorChannel`
 * (./space/mirror-channel.ts) is a second, independent implementation that
 * writes into a Starfish space instead.
 */
import type { SyncTrigger } from "./config.js"

export type { SyncTrigger }

/**
 * Passed into `ReplicaChannel.sync()` on every invocation.
 *
 * `callKind` distinguishes a replication-driven call (scheduled / on_pull /
 * syncNow / syncAll, always `"replicator"`) from a direct app call a channel
 * implementation may also expose (`"classic"`) — so one shared data-access
 * function on the app side can serve both call sites and branch on which one
 * this is, without the channel/manager needing to know why.
 */
export interface ReplicaCallContext {
  callKind: "replicator" | "classic"
  /** Optional cancellation signal a caller may pass to a long-running sync. */
  signal?: AbortSignal
}

/** The context `ReplicaManager` passes on every scheduling-driven call. */
export const REPLICATOR_CTX: ReplicaCallContext = { callKind: "replicator" }

/**
 * One collection's sync mechanics. `name` is the route key `ReplicaManager`
 * looks channels up by (matches `RemoteCollection.name` for the HTTP path).
 */
export interface ReplicaChannel {
  readonly name: string
  sync(ctx: ReplicaCallContext): Promise<void>
}

/** When/how often `ReplicaManager` drives a channel's `sync()`. */
export interface ChannelSchedule {
  /** Interval in ms for the `"scheduled"` trigger. Ignored otherwise. */
  intervalMs?: number
  /** Which events trigger a sync. */
  triggers: SyncTrigger[]
  /** Minimum ms between two `on_pull`-triggered syncs (cooldown). */
  onPullMinIntervalMs?: number
}

/** A channel plus the schedule `ReplicaManager` should drive it on. */
export interface ScheduledChannel {
  channel: ReplicaChannel
  schedule: ChannelSchedule
}
